core/ValidationModel.js

```diff
@@ -0,0 +1,143 @@
+
+/**
+ * Monitors fields in a model and updates the validation state when they change
+ */
+@HoistModel()
+export class ValidationModel {
+    /** The rules for validating each field */
+    constraints;
+
+    /** The model we are validating */
+    model;
+
+    /** Whether the current validation state is valid or not */
+    @observable isValid;
+
+    /** The current validation errors */
+    @observable.ref errors;
+
+    /**
+     * @param params
+     * @param params.constraints {Object} The validate.js constraints configuration object. @see https://validatejs.org/#constraints
+     * @param params.model {Object} The model to validate
+     */
+    constructor({constraints, model}) {
+        this.constraints = constraints;
+        this.model = model;
+
+        // Add reactions to validate individual fields on the model when they change
+        forOwn(constraints, (fieldConstraint, field) => {
+            // It is important to use hasIn here instead of has because the validation model will
+            // often be instantiated during construction of the model passed in here, so has or
+            // hasOwnProperty will fail most of the time. hasIn or the in operator will return true
+            // since the fields exist on the prototype.
+            if (!hasIn(model, field)) {
+                console.warn(
+                    `ValidationModel has constraint for field '${field}' which is not present in the model: `,
+                    model,
+                    'constraint: ',
+                    fieldConstraint
+                );
+                return;
+            }
+
+            this.addReaction({
+                track: () => model[field],
+                run: () => this.validateFieldAsync(field)
+            });
+        });
+    }
+
+    /**
+     * Checks whether a field is valid in the current validation state.
+     *
+     * @param field The field to check
+     * @returns {boolean} True if the field is valid, false otherwise
+     */
+    isFieldValid(field) {
+        return !has(this.errors, field);
+    }
+
+    /**
+     * Returns all validation error messages in a flat list
+     * @returns {Array} array of error message strings
+     */
+    get allErrorMessages() {
+        return flatten(values(this.errors));
+    }
+
+    /**
+     * Returns all validation error messages for the specified fields
+     * @param fields {Array|String} the fields to get the error messages for
+     * @returns {Array} array of error message string
+     */
+    getFieldErrors(fields) {
+        const {errors} = this;
+        if (!errors) return [];
+
+        fields = castArray(fields);
+        return flatten(fields.map(field => errors[field] || []));
+    }
+
+    /**
+     * Validates all fields on the model which have constraints configured.
+     *
+     * @returns {Promise} A Promise that resolves to an object containing the current validation state
+     */
+    async validateAsync() {
```

Comment:
@lbwexler John and I had discussed that the async version of this method seemed more powerful and consistent with our use of promise-based APIs elsewhere in the framework. It does introduce some possible complexities, however - e.g. seems like out-of-order returns could be a possible source of bugs here if we had truly async server-side validation.

The fact that the validation library rejects the promise on a validation failure also seemed like a bit of an odd choice to me (on their part).

Not sure if we need to consider or do anything with this now. I didn't / don't want to over-design.